### ISSUE TITLE
Show Linux processes in Task manager 

### ISSUE THREAD

**Issue by mikemaccana** [microsoft/WSL#6881]
**Is your feature request related to a problem? Please describe.**

Sometime my computer is slow, and I'd like to be able to see what's slowing it down. If a Linux process is slowing my machine down or using too many resources, I won't be able to see it in task manager. Instead I have to run an additional process inside the Linux VM (which may not even execute in a reasonable amount of time if the Linux VM is busy) to see what the process is.

**Describe the solution you'd like**

Microsoft Edge already reports usage of individual tabs to Task Manager:
![image](https://user-images.githubusercontent.com/172594/116546478-907eaf80-a8e9-11eb-908d-104a34afbb12.png)

It would be useful for the agent that runs in WSL2 to report the same details, allowing individual WSL2 processes to show up similar to how Edge tabs are.

Eg, say  if my 'next.js' build process has a memory leak, I should be able to find and stop the process from Task Manager.

[WSL1 already showed Linux processes](https://twitter.com/richturn_ms/status/1387469816680763400?s=20), according to Rich Turner, so this could be looked at in some ways as a regression (though obviously WSL2 and WSL1 have very different architectures).

**Describe alternatives you've considered**

Right now I just have to see it's nowt Windows in Task Manager, jump into the Linux VM and wait  for `htop` ot `top` or similar to start

**Additional context**

None. Thanks for WSL2 thought it's great! 🙂

**Comment by drkvogel** [comment]
Maybe this would be nice to have, but it's easy to find out what is using CPU in WSL. 

In the Windows Task Manager, if your distro in WSL is using a lot of CPU in general, you will see `vmmem.exe` using a lot of CPU - that is basically WSL.

In your distro in WSL, you can use `top` (built-in to most Linux distros), or https://htop.dev/ (which I prefer) to see what is using CPU within WSL.

**Comment by mikemaccana** [comment]
@drkvogel Yep, Task Manager will tell me that the Linux VM is using resources, but it won't tell me what the specific process is, or let me kill it.

**Comment by t9t** [comment]
Searching for how to view WSL2 processes in the Windows Task Manager I found this issue related to WSL1: #122

There is was explained that it could also be a security concern to not show WSL processes in the Task Manager, as bad actors could basically "hide" processes away from the user, running them inside WSL. The user will have to check both Task Manager (or Resource Monitor) and something like `top` inside WSL to see all the processes.

Further there is also this discussion going on already: #5828.

**Comment by JasonGross** [comment]
The security concern from WSL1 is not present here; it's quite obvious when a Linux process is consuming CPU, as it will show up as Vmmem

**Comment by WSLUser** [comment]
In WSL1, individual processes are shown in Task Manager and are able to be killed from there. They just need to adapt the code to account for WSL2, which is likely far more difficult than it sounds due to the major arch differences between 1 and 2.

**Comment by cerebrate** [comment]
I suspect doing this well might mean adding a "distro" column to task manager, such that you could tell which instance of certain processes was which - especially since a transparent way of doing this would have to include in the WSLg system distro processes and more. (After all, the obvious and transparent implementation would just be to pull the process list from the hidden namespace-above-all-namespaces.)

Or, although this is moving away from WSL to general Task Manager improvements, doing what I've long wanted and moving to a tree-structured display of processes and parents, like Process Explorer:

![image](https://user-images.githubusercontent.com/371623/117043169-81757400-acd2-11eb-8cc0-2a069d7b9a8c.png)

**Comment by souradeep100** [comment]
If I am not wrong, the process details get updated in linux kernel in in memory proc file system. Now exposing that to
root partition VM, will cause a memory security issue. As each VM should not see other VM's memory details.
So though it can be done, but it is against  the  memory opaqueness of virtualization. Just my thought here.

**Comment by cerebrate** [comment]
Individual distros don't run in their own VMs; they run in Linux namespaces (i.e., containers) underneath WSL's internal userspace. The way Linux pid namespaces are implemented, all processes in a namespace are visible to the parent namespace, just not vice versa. You can confirm this for yourself on a Linux machine running, for example, containerd or docker, on which `ps aux` will show you the processes within the containers as well as those without.

**Comment by souradeep100** [comment]
So all the wsl distros are like containers and they are using same WSL VM, but the WSL VM (guest )and Windows VM (host) are having two different virtual memory.
In that case from windows VM we can not access the /proc/pid of the WSL VM. Please let me know if I am missing something.

**Comment by souradeep100** [comment]
Just now installed kali and running both in windows.
But the proc file systems are different.

![image](https://user-images.githubusercontent.com/11392837/117200579-3e69ec80-ae09-11eb-9287-037a7b50ffce.png)

**Comment by cerebrate** [comment]
Bear in mind that there's already a way to access Linux files from Windows, in the form of the 9p file system that lets you see _//wsl.localhost/<distro>_ in Explorer. At the moment, I don't believe it works for all Linux filesystem types, /proc included, but the principle can be considered established.

**Comment by ANR2ME** [comment]
It would also be nice if we can distinguish which distro is a `vmmem` belongs to from Task manager.

Because currently i only have one distro (Ubuntu) and in Stopped state after `wsl --shutdown`, but i'm seeing 2 `vmmem`in Task manager O.o
One of them is in suspended state (which is the Ubuntu) and uses very small memory, while the other `vmmem` is running and using 500mb memory (peaked to nearly 1.5gb), and i don't know who the owner of this running `vmmem` and why it uses so much memory :(

**Comment by onomatopellan** [comment]
@ANR2ME In Windows 11 there is now the process `vmmemWSL` that is executed when a distro is running. If you only see `vmmem` then that's not WSL but other Hyper-V subset like Windows Sandbox, Virtual Based Security or even a Hyper-V VM.

**Comment by ANR2ME** [comment]
@onomatopellan I'm using Windows 11 21H2 (build 22000.593) but i never saw that `vmmemWSL`, are they only available on Insider Preview?

And i was mistaken about the `vmmem` in suspended state, apparently it's not the stopped Ubuntu (not sure what it was but since it doesn't use much resource i didn't really care)

This is after i restart Windows (`wsl -l -v` showed that Ubuntu is in Stopped state):
![image](https://user-images.githubusercontent.com/7974720/162674849-099d1e4e-9634-432e-96b2-7f712d9fa3e5.png)

This is when i ran Ubuntu terminal (but not running anything on linux), but no `vmmemWSL` in it:
![image](https://user-images.githubusercontent.com/7974720/162675241-02ae31fa-07db-4b9c-be5b-8ba3a85f67b7.png)
Couldn't even tell whether it was Ubuntu or not just based on Task Manager (may be a different distro will have different GUID on the Username?)

This is after closing Ubuntu terminal and doing `wsl --shutdown`, also making sure Ubuntu state has Stopped on `wsl -l -v`:
![image](https://user-images.githubusercontent.com/7974720/162675731-17924454-5d01-4ec1-be38-3d1d9ee1fa26.png)
Which is similar to the condition after restarting windows.

I can get rid of that running `vmmem` by killing the `vmwp.exe` with the same Username (GUID), but not sure what it was or whether it's safe to do or not, but i didn't experience any strange behavior after killing it.

Is there anyway to find out what kind of VM it was just based on the GUID? because i don't like have something i don't recognize taking up too much resource (was getting OutOfMemory issue pretty often while working with VS2019 if i had that `vmmem` running)

**Comment by onomatopellan** [comment]
@ANR2ME I'm in same windows 11 but you need to install the latest WSL [from the store](https://aka.ms/wslstorepage). After that you should see `vmmemWSL`.
You can also run `hcsdiag.exe -list` from elevated Powershell and it will tell you more about each GUID.

**Comment by ANR2ME** [comment]
> hcsdiag.exe list

Thanks, that `hcsdiag.exe list` helped in solving my previous issue, where that mysterious VM apparently related to Microsoft Defender Application Guard.

But today i'm facing another mysterious VM and the GUID doesn't appeared on `hcsdiag.exe list`
![image](https://user-images.githubusercontent.com/7974720/169258361-c18e4594-412d-490f-bd3c-fdc6583ded34.png)
I noticed that a VM is suddenly running by itself (because i usually ran `wsl --shutdown` after i'm done using Linux) and showing this RemoteApp dialog
![image](https://user-images.githubusercontent.com/7974720/169258774-5251b38a-9eff-48cc-a3d8-12100a009617.png)
And according to Task Manager, RemoteApp is part of WSL
![image](https://user-images.githubusercontent.com/7974720/169259115-0d0a52a3-80f9-4cab-9ee1-be5e88f1cf7b.png)

So i wanted to know whether it's safe to allow RemoteApp to connect? because this is my first time seeing such dialog.
